src/client/src/StreamState.c

```diff
@@ -477,13 +477,32 @@ STATUS executeGetEndpointStreamState(UINT64 customData, UINT64 time)
     STATUS retStatus = STATUS_SUCCESS;
     PKinesisVideoStream pKinesisVideoStream = STREAM_FROM_CUSTOM_DATA(customData);
     PKinesisVideoClient pKinesisVideoClient = NULL;
+    PStateMachineState pState;
+
 
     CHK(pKinesisVideoStream != NULL, STATUS_NULL_ARG);
 
     pKinesisVideoClient = pKinesisVideoStream->pKinesisVideoClient;
 
     // Step the client state machine first
-    CHK_STATUS(stepClientStateMachine(pKinesisVideoClient));
+    //CHK_STATUS(stepClientStateMachine(pKinesisVideoClient));
+
+    if(STATUS_FAILED(retStatus = stepClientStateMachine(pKinesisVideoClient))) {
+        if(retStatus == STATUS_CLIENT_AUTH_CALL_FAILED) {
+            // reset client state machine to READY state
```

Comment:
Not sure if it makes sense. You are getting the ready state and then stepping the state machine. There is a setState API for the state machine if you need it.

However, I would recommend incapsulating this within the state machine transition of the client. The fromBlahState API should check the result (which is STATUS_CLIENT_AUTH_CALL_FAILED and not transition to other state?

Reply:
Thanks for the comment @MushMal.  Here is the issue so when there have been no frames for a while and the network is bad the code sequence that goes through is this:

First `putFrame` call when it gets to this line:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/Stream.c#L781

It inside there, it sets this boolean to TRUE:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/Stream.c#L2736

Regardless of the result of `streamTerminatedEvent`

This executes this section of code:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/StreamEvent.c#L750-L765

This is stepping the STREAM state machine.  This will call now `fromStoppedStreamState` which sets the state to `STREAM_STATE_GET_ENDPOINT`.  Next we will call `executeGetEndpointStreamState`.  This method before my change will step the CLIENT state machine which as the log statements from the issue indicate will start in state `CLIENT_STATE_READY` and go to `CLIENT_STATE_AUTH`. Hence first we will invoke `fromAuthClientState` which will simply set state to `CLIENT_STATE_AUTH` next we invoke `executeAuthClientState`.  This will now FAIL here at `getAuthInfo`, here: https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/ClientState.c#L165

In producer C it fails and returns `STATUS_IOT_FAILED`:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/blob/a8091d1a140fbdcb260ba3c8302aa2e4712d2d91/src/source/Common/Curl/CurlCall.c#L80

Ok so now tracing this all the way up the call stack this is the error message returned by `putFrame`

Ok cool so this is fine... now the client calls `putFrame` again... NOW when we get here:

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/Stream.c#L781

We have a problem, we will just exit here:  https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/Stream.c#L2720

Because this boolean was never reset.  Notice though if I only make the change to flip the flag back to FALSE if `streamTerminatedEvent` returns that error then we have another problem.

Now we will step the stream state machine again but notice from here:  

https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/client/src/StreamEvent.c#L750-L765


It's always going to take the STREAM state machine to the `STREAM_STATE_STOPPED` which is good because the since the state is different from the state it came from `STREAM_STATE_GET_ENDPOINT` the retryCount will be reset back to zero so even on subsequent `putFrame` calls it'll never just stop doing its job.  However as part of this call when we get to the `executeGetEndpointStreamState` call we need to step the CLIENT state machine. There is the problem I'm trying to overcome this one is now in `CLIENT_STATE_AUTH` and it's stuck in that state there's nothing to reset it, so after 5 `putFrame` calls the max retry will be reached and now it'll return `STATUS_CLIENT_AUTH_CALL_FAILED`.  The next time `putFrame` is called then inside the `stepStateMachine` function it'll just exit early here:  https://github.com/awslabs/amazon-kinesis-video-streams-pic/blob/master/src/state/src/State.c#L170 and never get to the part of executing the state function.

This is what I want to avoid, because say after 100 `putFrame` calls the network is back up, on the next `putFrame` call it should go through but in this case the client state machine is stuck as we see in the Issue raised here:  https://github.com/awslabs/amazon-kinesis-video-streams-producer-c/issues/221